python/lsst/obs/decam/decamCpIsr.py

```diff
@@ -0,0 +1,145 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+from lsst.ip.isr import biasCorrection, flatCorrection
+from lsst.meas.algorithms.detection import SourceDetectionTask
+import lsst.pipe.base as pipeBase
+from .isr import DecamIsrTask
+
+
+def _computeEdgeSize(rawExposure, calibExposure):
+    """Compute the number of edge trim pixels of the calibration product.
+    Some Community Pipeline Calibration products are trimmed on their edges
+    and are smaller than the raw data. Use the dimension difference between
+    raw exposure and the calibration product to compute the edge trim pixels.
+    @param[in] rawExposure: the data section of a raw exposure
+    @param[in] calibExposure: calibration bias or flat exposure,
+                              known to be smaller than raw data
+    @return an integer as the number of trimmed pixels on each edge
+    """
+    nx, ny = rawExposure.getBBox().getDimensions() - calibExposure.getBBox().getDimensions()
+    assert nx == ny, "Exposure is trimmed differently in X and Y"
+    assert nx%2 == 0, "Exposure is trimmed unevenly in X"
+    assert nx >= 0, "Calibration image is larger than raw data"
+    return nx/2
+
+
+class DecamCpIsrTask(DecamIsrTask):
+    """Perform ISR task using Community Pipeline Calibration Products MasterCal
+
+    The CP MasterCal products have butler dataset types cpBias and cpFlat,
+    different from the LSST-generated calibration products (bias/flat).
+    """
+    def readIsrData(self, dataRef, rawExposure):
+        """Retrieve necessary frames for instrument signature removal
+        \param[in] dataRef -- a daf.persistence.butlerSubset.ButlerDataRef
+                              of the detector data to be processed
+        \param[in] rawExposure -- a reference raw exposure that will later be
+                                  corrected with the retrieved calibration data;
+                                  should not be modified in this method.
+        \return a pipeBase.Struct with fields containing kwargs expected by run()
+         - bias: exposure of bias frame
+         - dark: exposure of dark frame
+         - flat: exposure of flat field
+         - defects: list of detects
+         - fringeStruct: a pipeBase.Struct with field fringes containing
+                         exposure of fringe frame or list of fringe exposure
```

Comment:
The fringe part is fixed now. 

Because the CP fringe products are single HDU fits files too, unlike the CP bias or flat, we don't have to add another dataset type to support CP fringes and LSST fringes. (I wasn't sure about that back then.)  Now fringe is left out from this commit, and no longer need to duplicate the fringe codes to support fringe correction with CP products. As for the nearly identical `readIsrData`, I don't see a way to avoid that either other than refactoring IsrTask itself.